lint/variables.go

```diff
@@ -0,0 +1,164 @@
+				return val.Format(time.RFC3339)
+			default:
+				return "Unsupported momentjs time format: " + format
+			}
+		default:
+			switch format {
+			case "date":
+				return val.Format(time.RFC3339)
+			default:
+				return strconv.FormatInt(val.UnixMilli(), 10)
+			}
+		}
+	default:
+		svalue := fmt.Sprintf("%s", value)
+		svalueList := []string{svalue, svalue, svalue}
+		// Implements https://grafana.com/docs/grafana/latest/variables/advanced-variable-format-options/
+		switch format {
+		case "csv":
+			return strings.Join(svalueList, ",")
+		case "doublequote":
+			return "\"" + strings.Join(svalueList, "\",\"") + "\""
+		case "glob":
+			return "{" + strings.Join(svalueList, ",") + "}"
+		case "json":
+			data, err := json.Marshal(svalueList)
+			if err != nil {
+				return err.Error()
+			}
+			return string(data)
+		case "lucene":
+			return "(\"" + strings.Join(svalueList, "\" OR \"") + "\")"
+		case "percentencode":
+			return url.QueryEscape(strings.Join(svalueList, ","))
+		case "pipe":
+			return strings.Join(svalueList, "|")
+		case "raw":
+			return strings.Join(svalueList, ",")
+		case "regex":
+			return strings.Join(svalueList, "|")
+		case "singlequote":
+			return "'" + strings.Join(svalueList, "','") + "'"
+		case "sqlstring":
+			return "'" + strings.Join(svalueList, "','") + "'"
+		case "text":
+			return strings.Join(svalueList, " + ")
+		case "queryparam":
+			values := url.Values{}
+			for _, svalue := range svalueList {
+				values.Add("var-"+name, svalue)
+			}
+			return values.Encode()
+		default:
+			return svalue
+		}
+	}
+}
+
+func variableSampleValue(s string) string {
+	var name, kind, format string
+	parts := strings.Split(s, ":")
+	switch len(parts) {
+	case 1:
+		// No format
+		name = s
+	case 2:
+		// Could be __from:date, variable:csv, ...
+		name = parts[0]
+		format = parts[1]
+	case 3:
+		// Could be __from:date:iso, ...
+		name = parts[0]
+		kind = parts[1]
+		format = parts[2]
+	}
+	// If it is part of the globals, return a string representation of a sample value
+	if value, ok := globalVariables[name]; ok {
+		return stringValue(name, value, kind, format)
+	}
+	// Assume variable type is a string
+	return stringValue(name, name, kind, format)
+}
+
+var variableMatchers = []struct {
+	pattern *regexp.Regexp
+	repl    func(string) string
+}{
+	{
+		regexp.MustCompile(`\$[[:word:]]+`),
+		func(s string) string {
+			return fmt.Sprintf("%s", variableSampleValue(strings.TrimLeft(s, "$")))
+		},
+	},
+	{
+		regexp.MustCompile(`\$\{[^}]+\}`),
+		func(s string) string {
+			return fmt.Sprintf("%s", variableSampleValue(strings.Trim(s, "${}")))
+		},
+	},
+	{
+		regexp.MustCompile(`\[\[[^\[\]]+\]\]`),
+		func(s string) string {
+			return fmt.Sprintf("%s", variableSampleValue(strings.Trim(s, "[]")))
+		},
+	},
+}
+
+func expandVariables(expr string) string {
+	parts := strings.Split(expr, "\"")
+	for i := range parts {
+		if i%2 == 1 {
+			// Inside a double quote string, skip it
+			continue
+		}
```

Comment:
For the sake of simplicity, is this necessary?  I know PromQL will parse `{job="$job"}` but why not more closely mirror what Grafana does and also replace those inside quotes?

Reply:
Indeed, just confirmed that I get the error: `Message: (string) (len=144) "Dashboard 'dashboard', panel 'panel' invalid PromQL query 'sum(rate(foo{job=~\"$job\",instance=~\"$instance\"}[5m]))': job selector is job, not $job"`

Reply:
> think it will break the "panel-job-instance-rule" linter that checks that the variable used is correctly $job and $instance.

Ah of course, good point.